File: 0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int curSum = nums[0];
        int maxSum = curSum;

        for (int i = 1; i < nums.size(); ++i) {
            if (curSum < 0) {
                curSum = nums[i];
            } else {
                curSum += nums[i];
            }
            maxSum = max(maxSum, curSum);
        }
        return maxSum;
    }
};

File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # kandane algo
        # o(n)
        max_sum = -float("inf")
        cur_sum = 0

        for num in nums:
            cur_sum = max(0, cur_sum)
            cur_sum += num

            max_sum = max(cur_sum, max_sum)
        
        return max_sum
        